pqarrow/frostdb: accumulate results in Iterator

asubiotto opened the pull request:
Previously, in some cases, the Table.Iterator method would push Records with a
single row one at a time to the next operator. This would happen especially in
cases where a distinct optimization is executed and a single value is written
per parquet page. This commit changes the scan process to write scan results to
a provided RecordBuilder in order to buffer results up to a specified
builderBufferSize (default=1024) and take advantage of processing these results
in tight loops downstream.

The results are especially noticeable in QueryTypes, which takes advantage of
the aforementioned distinct optimizations:

```
name          old time/op    new time/op    delta
QueryTypes-8     169ms ±11%      68ms ±13%  -59.43%  (p=0.000 n=10+9)
QueryMerge-8    27.0ms ±34%    23.4ms ±12%     ~     (p=0.095 n=10+9)
QueryRange-8    13.8ms ±15%    13.3ms ± 9%     ~     (p=0.356 n=9+10)

name          old alloc/op   new alloc/op   delta
QueryTypes-8     443MB ± 0%      88MB ± 0%  -80.21%  (p=0.000 n=10+9)
QueryMerge-8    39.9MB ± 0%    39.9MB ± 0%     ~     (p=0.497 n=10+9)
QueryRange-8    18.6MB ± 0%    18.6MB ± 0%   -0.05%  (p=0.000 n=10+10)

name          old allocs/op  new allocs/op  delta
QueryTypes-8     1.13M ± 0%     0.52M ± 0%  -53.55%  (p=0.000 n=10+9)
QueryMerge-8      104k ± 0%      104k ± 0%   -0.10%  (p=0.000 n=10+9)
QueryRange-8     6.61k ± 0%     6.49k ± 0%   -1.69%  (p=0.000 n=8+10)
```

asubiotto commented:
I think I would still like to add a test or two (even if the test suite caught a bunch of things). I'll leave exploring improving the intermediate result cache on top of the record builder (to avoid having to recreate a builder if an optimization cannot be applied) to a future PR.

asubiotto commented:
Cherry-picking #164 on to this PR gives us an addition 10% improvement:
```
name          old time/op    new time/op    delta
QueryTypes-8     169ms ±11%      50ms ±13%  -70.19%  (p=0.030 n=10+2)

name          old alloc/op   new alloc/op   delta
QueryTypes-8     443MB ± 0%      81MB ± 0%  -81.80%  (p=0.030 n=10+2)

name          old allocs/op  new allocs/op  delta
QueryTypes-8     1.13M ± 0%     0.45M ± 0%  -60.34%  (p=0.000 n=10+2)
```

thorfour reviewed (approved):
![](https://media.giphy.com/media/XjEvy8X8XB6Yo/giphy.gif)

Love this change

asubiotto commented:
Added a test for list types and refactored the list building code while I was at it. I'll let this PR sit for a bit in case there are further comments, but planning on merging this soon.
